# PIFuHD Demo: https://shunsukesaito.github.io/PIFuHD/

![](https://shunsukesaito.github.io/PIFuHD/resources/images/pifuhd.gif)

Made by [![Follow](https://img.shields.io/twitter/follow/psyth91?style=social)](https://twitter.com/psyth91)

To see how the model works, visit the project repository.

[![GitHub stars](https://img.shields.io/github/stars/facebookresearch/pifuhd?style=social)](https://github.com/facebookresearch/pifuhd)

## Note
Make sure that your runtime type is 'Python 3 with GPU acceleration'. To do so, go to Edit > Notebook settings > Hardware Accelerator > Select "GPU".

## More Info
- Paper: https://arxiv.org/pdf/2004.00452.pdf
- Repo: https://github.com/facebookresearch/pifuhd
- Project Page: https://shunsukesaito.github.io/PIFuHD/
- 1-minute/5-minute Presentation (see below)

In [ ]:
import IPython
IPython.display.HTML('<h2>1-Minute Presentation</h2><iframe width="720" height="405" src="https://www.youtube.com/embed/-1XYTmm8HhE" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe><br><h2>5-Minute Presentation</h2><iframe width="720" height="405" src="https://www.youtube.com/embed/uEDqCxvF5yc" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>')

## Requirements
- Python 3
- PyTorch tested on 1.4.0
- json
- PIL
- skimage
- tqdm
- numpy
- cv2

## Help! I'm new to Google Colab

You can check out the following youtube video on how to upload your own picture and run PIFuHD. **Note that with new update, you can upload your own picture more easily with GUI down below.**


In [ ]:
import IPython
IPython.display.HTML('<iframe width="720" height="405" src="https://www.youtube.com/embed/LWDGR5v3-3o" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>')

## Clone PIFuHD repository

In [ ]:
!git clone https://github.com/facebookresearch/pifuhd

Cloning into 'pifuhd'...
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 195 (delta 0), reused 1 (delta 0), pack-reused 191
Receiving objects: 100% (195/195), 399.23 KiB | 14.26 MiB/s, done.
Resolving deltas: 100% (93/93), done.


### install requirement.txt 하는법


In [101]:
cd pifuhd

d:\2023\3d\pifuhd


In [ ]:
!pip install -r requirements.txt

## Configure input data

In [103]:
cd ..

d:\2023\3d


In [104]:
cd ./pifuhd/sample_images

d:\2023\3d\pifuhd\sample_images


**If you want to upload your own picture, run the next cell**. Otherwise, go to the next next cell. Currently PNG, JPEG files are supported.

In [ ]:
# from google.colab import files

# filename = list(files.upload().keys())[0]

IndexError: ignored

In [105]:
import os

try:
  image_path = './pifuhd/sample_images/%s' % filename
except:
  image_path = './pifuhd/sample_images/test.png' # example image
image_dir = os.path.dirname(image_path)
file_name = os.path.splitext(os.path.basename(image_path))[0]

# output pathes
obj_path = './pifuhd/results/pifuhd_final/recon/result_%s_256.obj' % file_name
out_img_path = './pifuhd/results/pifuhd_final/recon/result_%s_256.png' % file_name
video_path = './pifuhd/results/pifuhd_final/recon/result_%s_256.mp4' % file_name
video_display_path = './pifuhd/results/pifuhd_final/result_%s_256_display.mp4' % file_name

In [106]:
print(image_path)

./pifuhd/sample_images/test.png


In [107]:
os.getcwd()

'd:\\2023\\3d\\pifuhd\\sample_images'

In [108]:
cd ..

d:\2023\3d\pifuhd


In [109]:
cd ..

d:\2023\3d


## Preprocess (for cropping image)

#### pose estimation 라이브러리를 깃허브에서 받고
#### pifuhd폴더와 같은 레벨에 위치시킨다.

In [13]:
!git clone https://github.com/Daniil-Osokin/lightweight-human-pose-estimation.pytorch.git

Cloning into 'lightweight-human-pose-estimation.pytorch'...


In [110]:
cd ./lightweight-human-pose-estimation.pytorch/

d:\2023\3d\lightweight-human-pose-estimation.pytorch


### 1. 이거 윈도우에서는 wget 명령어가 실행이 안되는데 그냥 다운로드 받아서 업로드 하면 됨 혹은
### 2. wget윈도우 사용방법 구글링해서 하면 나옴
### wget 옆에 문자 꺠진건 한글로 실행 안된다는 내용
### 지금 이 파일에서는 직접 다운받아서 폴더에 집어넣었어요

In [15]:
!wget https://download.01.org/opencv/openvino_training_extensions/models/human_pose_estimation/checkpoint_iter_370000.pth

'wget'��(��) ���� �Ǵ� �ܺ� ����, ������ �� �ִ� ���α׷�, �Ǵ�
��ġ ������ �ƴմϴ�.


In [111]:
os.getcwd()


'd:\\2023\\3d\\lightweight-human-pose-estimation.pytorch'

In [48]:
!pip install pycocotools

  Using cached pycocotools-2.0.6-cp39-cp39-win_amd64.whl
     ---------------------------------------- 0.0/7.6 MB ? eta -:--:--
     ---------------------------------------- 0.0/7.6 MB ? eta -:--:--
     ---------------------------------------- 0.0/7.6 MB 660.6 kB/s eta 0:00:12
      --------------------------------------- 0.1/7.6 MB 1.1 MB/s eta 0:00:07
     ------ --------------------------------- 1.2/7.6 MB 8.1 MB/s eta 0:00:01
     ------------- -------------------------- 2.5/7.6 MB 13.4 MB/s eta 0:00:01
     ---------------------- ----------------- 4.2/7.6 MB 17.9 MB/s eta 0:00:01
     ------------------------------- -------- 6.0/7.6 MB 21.4 MB/s eta 0:00:01
     ---------------------------------------  7.6/7.6 MB 23.2 MB/s eta 0:00:01
     ---------------------------------------- 7.6/7.6 MB 22.1 MB/s eta 0:00:00
  Using cached importlib_resources-5.12.0-py3-none-any.whl (36 kB)
  Using cached cycler-0.11.0-py3-none-any.whl (6.4 kB)
     ---------------------------------------- 0.

In [112]:
import torch
import cv2
import numpy as np
from models.with_mobilenet import PoseEstimationWithMobileNet
from modules.keypoints import extract_keypoints, group_keypoints
from modules.load_state import load_state
from modules.pose import Pose, track_poses
import demo

def get_rect(net, images, height_size):
    net = net.eval()

    stride = 8
    upsample_ratio = 4
    num_keypoints = Pose.num_kpts
    previous_poses = []
    delay = 33
    for image in images:
        rect_path = image.replace('.%s' % (image.split('.')[-1]), '_rect.txt')
        img = cv2.imread(image, cv2.IMREAD_COLOR)
        orig_img = img.copy()
        orig_img = img.copy()
        heatmaps, pafs, scale, pad = demo.infer_fast(net, img, height_size, stride, upsample_ratio, cpu=False)

        total_keypoints_num = 0
        all_keypoints_by_type = []
        for kpt_idx in range(num_keypoints):  # 19th for bg
            total_keypoints_num += extract_keypoints(heatmaps[:, :, kpt_idx], all_keypoints_by_type, total_keypoints_num)

        pose_entries, all_keypoints = group_keypoints(all_keypoints_by_type, pafs)
        for kpt_id in range(all_keypoints.shape[0]):
            all_keypoints[kpt_id, 0] = (all_keypoints[kpt_id, 0] * stride / upsample_ratio - pad[1]) / scale
            all_keypoints[kpt_id, 1] = (all_keypoints[kpt_id, 1] * stride / upsample_ratio - pad[0]) / scale
        current_poses = []

        rects = []
        for n in range(len(pose_entries)):
            if len(pose_entries[n]) == 0:
                continue
            pose_keypoints = np.ones((num_keypoints, 2), dtype=np.int32) * -1
            valid_keypoints = []
            for kpt_id in range(num_keypoints):
                if pose_entries[n][kpt_id] != -1.0:  # keypoint was found
                    pose_keypoints[kpt_id, 0] = int(all_keypoints[int(pose_entries[n][kpt_id]), 0])
                    pose_keypoints[kpt_id, 1] = int(all_keypoints[int(pose_entries[n][kpt_id]), 1])
                    valid_keypoints.append([pose_keypoints[kpt_id, 0], pose_keypoints[kpt_id, 1]])
            valid_keypoints = np.array(valid_keypoints)
            
            if pose_entries[n][10] != -1.0 or pose_entries[n][13] != -1.0:
              pmin = valid_keypoints.min(0)
              pmax = valid_keypoints.max(0)

              center = (0.5 * (pmax[:2] + pmin[:2])).astype(np.int64)
              radius = int(0.65 * max(pmax[0]-pmin[0], pmax[1]-pmin[1]))
            elif pose_entries[n][10] == -1.0 and pose_entries[n][13] == -1.0 and pose_entries[n][8] != -1.0 and pose_entries[n][11] != -1.0:
              # if leg is missing, use pelvis to get cropping
              center = (0.5 * (pose_keypoints[8] + pose_keypoints[11])).astype(np.int64)
              radius = int(1.45*np.sqrt(((center[None,:] - valid_keypoints)**2).sum(1)).max(0))
              center[1] += int(0.05*radius)
            else:
              center = np.array([img.shape[1]//2,img.shape[0]//2])
              radius = max(img.shape[1]//2,img.shape[0]//2)

            x1 = center[0] - radius
            y1 = center[1] - radius

            rects.append([x1, y1, 2*radius, 2*radius])

        np.savetxt(rect_path, np.array(rects), fmt='%d')

In [113]:
import os
os.getcwd()

'd:\\2023\\3d\\lightweight-human-pose-estimation.pytorch'

In [53]:
cd ../lightweight-human-pose-estimation.pytorch/

d:\2023\3d\lightweight-human-pose-estimation.pytorch


In [54]:
print(image_path)

./pifuhd/sample_images/test.png


In [39]:
import os 
os.getcwd()

'd:\\2023\\3d\\lightweight-human-pose-estimation.pytorch'

In [71]:
cd .. 

d:\2023\3d


In [72]:
net = PoseEstimationWithMobileNet()
checkpoint = torch.load('./lightweight-human-pose-estimation.pytorch/checkpoint_iter_370000.pth', map_location='cpu')
load_state(net, checkpoint)

get_rect(net.cuda(), [image_path], 512)

## Download the Pretrained Model

In [61]:
print(torch.cuda.is_available())

True


In [62]:
cd ./pifuhd

d:\2023\3d\pifuhd


In [44]:
!sh ./scripts/download_trained_model.sh

'sh'��(��) ���� �Ǵ� �ܺ� ����, ������ �� �ִ� ���α׷�, �Ǵ�
��ġ ������ �ƴմϴ�.


In [74]:
os.getcwd()

'd:\\2023\\3d'

## Run PIFuHD!


In [79]:
cd ..

d:\2023\3d\pifuhd


In [80]:
# Warning: all images with the corresponding rectangle files under -i will be processed. 
!python -m apps.simple_test -r 256 --use_rect

# seems that 256 is the maximum resolution that can fit into Google Colab. 
# If you want to reconstruct a higher-resolution mesh, please try with your own machine. 

Resuming from  ./checkpoints/pifuhd.pt
test data size:  1
initialize network with normal
initialize network with normal
generate mesh (test) ...
./results/pifuhd_final/recon/result_test_256.obj
module 'numpy' has no attribute 'bool'.
`np.bool` was a deprecated alias for the builtin `bool`. To avoid this error in existing code, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations



  0%|          | 0/1 [00:00<?, ?it/s]d:\2023\3d\pifuhd\lib\sdf.py:93: FutureWarning: In the future `np.bool` will be defined as the corresponding NumPy scalar.
  notprocessed = np.zeros(resolution, dtype=np.bool)

100%|██████████| 1/1 [00:08<00:00,  8.11s/it]


## Render the result

In [1]:
import torch
torch.__version__

'1.13.1'

In [2]:
!nvcc --version

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Tue_Mar__8_18:36:24_Pacific_Standard_Time_2022
Cuda compilation tools, release 11.6, V11.6.124
Build cuda_11.6.r11.6/compiler.31057947_0


In [55]:
!nvidia-smi

Tue Apr  4 00:53:22 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 528.49       Driver Version: 528.49       CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ... WDDM  | 00000000:23:00.0  On |                  N/A |
|  0%   43C    P8    22W / 200W |    899MiB /  8192MiB |     32%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

### pytorch3d가 특정 조건에서만 돌아가는데 윈도우는 바로 아래코드를 실행시키면 됨

In [81]:
!pip install torch torchvision torchaudio -f https://download.pytorch.org/whl/cu111/torch_stable.html
!pip install git+https://github.com/facebookresearch/pytorch3d.git@stable

Looking in links: https://download.pytorch.org/whl/cu111/torch_stable.html
  Cloning https://github.com/facebookresearch/pytorch3d.git (to revision stable) to c:\users\aivoi\appdata\local\temp\pip-req-build-fyb74dwd
  Resolved https://github.com/facebookresearch/pytorch3d.git to commit 3145dd4d16edaceb394838364b8e87a440f83c10
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Created wheel for pytorch3d: filename=pytorch3d-0.7.2-cp39-cp39-win_amd64.whl size=5529197 sha256=0ba39a1d9c9a96e9fccb427a829f7efc458e1dd01913c2991bc3932c11a87103
  Stored in directory: C:\Users\aivoi\AppData\Local\Temp\pip-ephem-wheel-cache-0fg3_jlc\wheels\d7\e2\25\51fb3f170ac9f43bf7d1bd4bdbf67e4a34104adc2ecfe5f62e
Successfully built pytorch3d


  Running command git clone --filter=blob:none --quiet https://github.com/facebookresearch/pytorch3d.git 'C:\Users\aivoi\AppData\Local\Temp\pip-req-build-fyb74dwd'
  Running command git checkout -q 3145dd4d16edaceb394838364b8e87a440f83c10


  Cloning https://github.com/facebookresearch/pytorch3d.git (to revision stable) to c:\users\aivoi\appdata\local\temp\pip-req-build-i911_czr
  Resolved https://github.com/facebookresearch/pytorch3d.git to commit 3145dd4d16edaceb394838364b8e87a440f83c10
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Created wheel for pytorch3d: filename=pytorch3d-0.7.2-cp39-cp39-win_amd64.whl size=5529207 sha256=dcbbc9ae0bbac7a1c2d7b5802b543535cb12626995361a40981c55682a4e1f56
  Stored in directory: C:\Users\aivoi\AppData\Local\Temp\pip-ephem-wheel-cache-ntiyyjf9\wheels\d7\e2\25\51fb3f170ac9f43bf7d1bd4bdbf67e4a34104adc2ecfe5f62e
Successfully built pytorch3d


  Running command git clone --filter=blob:none --quiet https://github.com/facebookresearch/pytorch3d.git 'C:\Users\aivoi\AppData\Local\Temp\pip-req-build-i911_czr'
  Running command git checkout -q 3145dd4d16edaceb394838364b8e87a440f83c10


### linux에서는 pytorch3d를 아래와 같은 코드를 돌리면 다운 받아짐

py39_cu116_pyt1131를 맞춰야함 cudnn은 cuda 11.6버전에 맞는 8.2.1~8.4.1중 아무거나

In [114]:
print(version_str)


py39_cu116_pyt1131


In [65]:
import sys
import torch
pyt_version_str=torch.__version__.split("+")[0].replace(".", "")
version_str="".join([
    f"py3{sys.version_info.minor}_cu",
    torch.version.cuda.replace(".",""),
    f"_pyt{pyt_version_str}"
])
!pip install fvcore iopath
!pip install --no-index --no-cache-dir pytorch3d -f https://dl.fbaipublicfiles.com/pytorch3d/packaging/wheels/{version_str}/download.html

  Using cached fvcore-0.1.5.post20221221-py3-none-any.whl
  Using cached iopath-0.1.10-py3-none-any.whl
  Using cached yacs-0.1.8-py3-none-any.whl (14 kB)
  Using cached termcolor-2.2.0-py3-none-any.whl (6.6 kB)
     ---------------------------------------- 0.0/151.6 kB ? eta -:--:--
     -- ------------------------------------- 10.2/151.6 kB ? eta -:--:--
     ------- ----------------------------- 30.7/151.6 kB 325.1 kB/s eta 0:00:01
     ----------------- ------------------- 71.7/151.6 kB 563.7 kB/s eta 0:00:01
     ------------------------------------ 151.6/151.6 kB 906.4 kB/s eta 0:00:00
  Using cached portalocker-2.7.0-py2.py3-none-any.whl (15 kB)
Looking in links: https://dl.fbaipublicfiles.com/pytorch3d/packaging/wheels/py39_cu116_pyt1131/download.html


ERROR: Could not find a version that satisfies the requirement pytorch3d (from versions: none)
ERROR: No matching distribution found for pytorch3d


In [82]:
from lib.colab_util import generate_video_from_obj, set_renderer, video

renderer = set_renderer()
generate_video_from_obj(obj_path, out_img_path, video_path, renderer)

# we cannot play a mp4 video generated by cv2
!ffmpeg -i $video_path -vcodec libx264 $video_display_path -y -loglevel quiet
video(video_display_path)

AttributeError: 'NoneType' object has no attribute 'shape'

## Tips for Inputs: My results are broken!

(Kudos to those who share results on twitter with [#pifuhd](https://twitter.com/search?q=%23pifuhd&src=recent_search_click&f=live) tag!!!!)

Due to the limited variation in the training data, your results might be broken sometimes. Here I share some useful tips to get resonable results. 

*   Use high-res image. The model is trained with 1024x1024 images. Use at least 512x512 with fine-details. Low-res images and JPEG artifacts may result in unsatisfactory results. 
*   Use an image with a single person. If the image contain multiple people, reconstruction quality is likely degraded.
*   Front facing with standing works best (or with fashion pose)
*   The entire body is covered within the image. (Note: now missing legs is partially supported)
*   Make sure the input image is well lit. Exteremy dark or bright image and strong shadow often create artifacts.
*   I recommend nearly parallel camera angle to the ground. High camera height may result in distorted legs or high heels. 
*   If the background is cluttered, use less complex background or try removing it using https://www.remove.bg/ before processing.
*   It's trained with human only. Anime characters may not work well (To my surprise, indeed many people tried it!!).
*   Search on twitter with [#pifuhd](https://twitter.com/search?q=%23pifuhd&src=recent_search_click&f=live) tag to get a better sense of what succeeds and what fails. 


## Share your result! 
Please share your results with[ #pifuhd](https://twitter.com/search?q=%23pifuhd&src=recent_search_click&f=live) tag on Twitter. Sharing your good/bad results helps and encourages the authors to further push towards producition-quality human digitization at home.
**As the tweet buttom below doesn't add the result video automatically, please download the result video above and manually add it to the tweet.**

In [ ]:
import IPython
IPython.display.HTML('<a href="https://twitter.com/intent/tweet?button_hashtag=pifuhd&ref_src=twsrc%5Etfw" class="twitter-hashtag-button" data-size="large" data-text="Google Colab Link: " data-url="https://bit.ly/37sfogZ" data-show-count="false">Tweet #pifuhd</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>  (Don\'t forget to add your result to the tweet!)')

## Cool Applications
Special thanks to those who play with PIFuHD and came up with many creative applications!! If you made any cool applications, please tweet your demo with [#pifuhd](https://twitter.com/search?q=%23pifuhd&src=recent_search_click&f=live). I'm constantly checking results there.
If you need complete texture on the mesh, please try my previous work [PIFu](https://github.com/shunsukesaito/PIFu) as well! It supports 3D reconstruction + texturing from a single image although the geometry quality may not be as good as PIFuHD.

In [ ]:
IPython.display.HTML('<h2>Rigging (Mixamo) + Photoreal Rendering (Blender)</h2><blockquote class="twitter-tweet"><p lang="pt" dir="ltr">vcs ainda tem a PACHORRA de me dizer que eu não sei dançar<a href="https://twitter.com/hashtag/b3d?src=hash&amp;ref_src=twsrc%5Etfw">#b3d</a> <a href="https://twitter.com/hashtag/pifuhd?src=hash&amp;ref_src=twsrc%5Etfw">#pifuhd</a> <a href="https://t.co/kHCnLh6zxH">pic.twitter.com/kHCnLh6zxH</a></p>&mdash; lukas arendero (@lukazvd) <a href="https://twitter.com/lukazvd/status/1274810484798128131?ref_src=twsrc%5Etfw">June 21, 2020</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script><h2>FaceApp + Rigging (Mixamo)</h2><blockquote class="twitter-tweet"><p lang="ja" dir="ltr">カツラかぶってる自分に見える <a href="https://twitter.com/hashtag/pifuhd?src=hash&amp;ref_src=twsrc%5Etfw">#pifuhd</a> <a href="https://t.co/V8o7VduTiG">pic.twitter.com/V8o7VduTiG</a></p>&mdash; Shuhei Tsuchida (@shuhei2306) <a href="https://twitter.com/shuhei2306/status/1274507242910314498?ref_src=twsrc%5Etfw">June 21, 2020</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script><h2>Rigging (Mixamo) + AR (Adobe Aero)</AR><blockquote class="twitter-tweet"><p lang="ja" dir="ltr">写真→PIFuHD→Mixamo→AdobeAeroでサウンド付きARを作成。Zip化してLINEでARコンテンツを共有。<br>写真が1枚あれば簡単にARの3Dアニメーションが作れる時代…凄い。<a href="https://twitter.com/hashtag/PIFuHD?src=hash&amp;ref_src=twsrc%5Etfw">#PIFuHD</a> <a href="https://twitter.com/hashtag/AdobeAero?src=hash&amp;ref_src=twsrc%5Etfw">#AdobeAero</a> <a href="https://twitter.com/hashtag/Mixamo?src=hash&amp;ref_src=twsrc%5Etfw">#Mixamo</a> <a href="https://t.co/CbiMi4gZ0K">pic.twitter.com/CbiMi4gZ0K</a></p>&mdash; モジョン (@mojon1) <a href="https://twitter.com/mojon1/status/1273217947872317441?ref_src=twsrc%5Etfw">June 17, 2020</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script><h2>3D Printing</h2><blockquote class="twitter-tweet"><p lang="ja" dir="ltr"><a href="https://twitter.com/hashtag/pifuhd?src=hash&amp;ref_src=twsrc%5Etfw">#pifuhd</a> 楽しい〜<br>小さい自分プリントした <a href="https://t.co/4qyWuij0Hs">pic.twitter.com/4qyWuij0Hs</a></p>&mdash; isb (@vxzxzxzxv) <a href="https://twitter.com/vxzxzxzxv/status/1273136266406694913?ref_src=twsrc%5Etfw">June 17, 2020</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>')